src/goto-symex/execution_state.cpp

```diff
@@ -727,6 +727,7 @@ unsigned int execution_statet::add_thread(const goto_programt *prog)
   unsigned int thread_nr = threads_state.size();
   new_state.source.thread_nr = thread_nr;
   new_state.global_guard = cur_state->guard;
+  new_state.guard = cur_state->guard;
```

Comment:
I've updated the global_guard one line above your addition in #1463, which had a similar aim to the one your PR describes, and it fixed a couple of SV-COMP benchmarks.

Since with this PR's change the cur_state->guard would be copied to both, the new_state's guard and also its global_guard, I'm wondering whether this duplication is necessary.

Reply:
I think the global guard changes might work in this case:
```
int main(void) {
    pthread_create(&t1, NULL, thread1, NULL); // global guard
    pthread_create(&t2, NULL, thread2, NULL); // global guard
    return 0;
}
```
The current guard is reflected in the guards in the current thread:
```
void *thread1(void *arg) {
    pthread_create(&t2, NULL, thread2, NULL);   // inherit the guard from thread1
    return 0;
}

int main(void) {
    pthread_create(&t1, NULL, thread1, NULL);
    return 0;
}
```
I did not understand this part deeply, please correct me if I am wrong.

Reply:
> This means that in the new thread, `cur guard` is empty, so `new_state.global_guard = cur_state->guard;` will not work when we create new thread in this child thread.

Can you explain why it wouldn't work?

Reply:
I never really got this code, what's the difference between the global guard and the state guard? Do we need both?

Reply:
My understanding is that the current guard is added to `global_guard` when we create a new thread.
https://github.com/esbmc/esbmc/blob/69fbe894f3733bb808d5d57d9d35fb9653a71070/src/goto-symex/builtin_functions.cpp#L683

Both the guard and the global guard of cur states are used when an assertion is executed.
https://github.com/esbmc/esbmc/blob/69fbe894f3733bb808d5d57d9d35fb9653a71070/src/goto-symex/symex_main.cpp#L104-L105